signer/signer.go

```diff
@@ -0,0 +1,261 @@
+package signer
+
+import (
+	"crypto/x509"
+	"fmt"
+	nx509 "github.com/notaryproject/notation-core-go/x509"
+	"time"
+)
+
+type SignatureMediaType string
+
+// SignatureAlgorithm lists supported Signature algorithms.
+type SignatureAlgorithm string
+
+const (
+	RSASSA_PSS_SHA_256 SignatureAlgorithm = "RSASSA_PSS_SHA_256"
+	RSASSA_PSS_SHA_384 SignatureAlgorithm = "RSASSA_PSS_SHA_384"
+	RSASSA_PSS_SHA_512 SignatureAlgorithm = "RSASSA_PSS_SHA_512"
+	ECDSA_SHA_256      SignatureAlgorithm = "ECDSA_SHA_256"
+	ECDSA_SHA_384      SignatureAlgorithm = "ECDSA_SHA_384"
+	ECDSA_SHA_512      SignatureAlgorithm = "ECDSA_SHA_512"
+)
+
+// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm SignatureAlgorithm
+	CertificateChain   []*x509.Certificate
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignedAttributes represents signed metadata in the Signature envelope
+type SignedAttributes struct {
+	SigningTime        time.Time
+	Expiry             time.Time
+	ExtendedAttributes []Attribute
+}
+
+// UnsignedAttributes represents unsigned metadata in the Signature envelope
+type UnsignedAttributes struct {
+	SigningAgent string
+}
+
+// SignRequest is used to generate Signature.
+type SignRequest struct {
+	Payload             []byte
+	PayloadContentType  string
+	CertificateChain    []*x509.Certificate
+	SignatureProvider   SignatureProvider
+	SigningTime         time.Time
+	Expiry              time.Time
+	ExtendedSignedAttrs []Attribute
+	SigningAgent        string
+}
+
+type Attribute struct {
+	Key      string
+	Critical bool
+	Value    interface{}
+}
+
+// SignatureProvider is used to sign bytes generated after creating Signature envelope.
+type SignatureProvider interface {
+	Sign([]byte) ([]byte, error)
+}
+
+type SignatureEnvelope struct {
```

Comment:
Why not make `SignatureEnvelope` an interface lile
```golang
type SignatureEnvelope interface {
    Sign(req SignRequest) ([]byte, error)
    Verify(trustedCerts []*x509.Certificate) (*x509.Certificate, *SignerInfo, error)
    GetSignerInfo() (*SignerInfo, error)
}
```

Then we can have stuffs simplified like

```go
// NewSignatureEnvelope is used for signature generation workflow
func NewSignatureEnvelope(envelopeMediaType SignatureMediaType) (SignatureEnvelope, error) {
	switch envelopeMediaType {
	case MediaTypeJWSJSON:
		return &jwsEnvelope{}, nil
	case MediaTypeCOSE:
		return &coseEnvelope{}, nil
	}
	return nil, &ErrorUnsupportedSignatureFormat{mediaType: string(envelopeMediaType)}
}
```

Reply:
So basically we want to implement common validations(like input validations, cert chain validations, trust anchor related code, timestamping related code) in SignatureEnvelope's sign and verify method and let specific implementation only do the work of encoding/decoding/integrity validation signature format.
With interface aboce things are not possible. Although there can be workaround by creating helper methods but than we need to make sure each implementation is calling them.

Reply:
Then we can still have

```go
// NewSignatureEnvelope is used for signature generation workflow
func NewSignatureEnvelope(envelopeMediaType SignatureMediaType) (SignatureEnvelope, error) {
	switch envelopeMediaType {
	case MediaTypeJWSJSON:
		return &someGenericValidation{&jwsEnvelope{}}, nil
	case MediaTypeCOSE:
		return &someGenericValidation{&coseEnvelope{}}, nil
	}
	return nil, &UnsupportedSignatureFormatError{mediaType: string(envelopeMediaType)}
}
```

That still works. With interfaces, it is easier to decouple code and do unit testing.